ngraph/core/src/op/interpolate.cpp

```diff
@@ -111,143 +114,538 @@ namespace ngraph
+        ++i;
     }
+}
 
-    if (auto const_shape = as_type_ptr<op::Constant>(input_value(1).get_node_shared_ptr()))
+void op::v4::Interpolate::infer_using_shapes(PartialShape& output_shape,
+                                             const std::vector<int64_t>& axes)
+{
+    auto const_shape = as_type_ptr<op::Constant>(input_value(1).get_node_shared_ptr());
+    if (!const_shape)
     {
-        auto out_shape = const_shape->cast_vector<int64_t>();
-        size_t i = 0;
-        for (auto axis : m_attrs.axes)
+        return;
+    }
+
+    auto out_shape = const_shape->cast_vector<int64_t>();
+    size_t i = 0;
+    for (auto axis : axes)
+    {
+        output_shape[axis] = Dimension(out_shape[i++]);
+    }
+}
+
+void op::v4::Interpolate::validate_and_infer_types()
+{
+    PartialShape output_shape = PartialShape(get_input_partial_shape(0));
+    PartialShape padded_input_shape = output_shape;
+
+    if (!output_shape.rank().is_static())
+    {
+        set_output_type(0, get_input_element_type(0), output_shape);
+        return;
+    }
+
+    auto axes = get_axes();
+    correct_pads();
+
+    const auto input_rank = output_shape.rank().get_length();
+
+    for (size_t i = 0; i < input_rank; ++i)
+    {
+        if (output_shape[i].is_static())
         {
-            output_shape[axis] = Dimension(out_shape[i++]);
+            auto new_length =
+                m_attrs.pads_begin[i] + m_attrs.pads_end[i] + output_shape[i].get_length();
+            output_shape[i] = Dimension(new_length);
+            padded_input_shape[i] = Dimension(new_length);
         }
     }
+
+    for (auto axis : axes)
+    {
+        NGRAPH_CHECK(axis < input_rank);
+        output_shape[axis] = Dimension::dynamic();
+    }
+
+    if (m_attrs.shape_calculation_mode == ShapeCalcMode::scales)
+    {
+        infer_using_scales(output_shape, axes, padded_input_shape);
+    }
+    else
+    {
+        infer_using_shapes(output_shape, axes);
+    }
+
     set_output_type(0, get_input_element_type(0), output_shape);
 }
 
-shared_ptr<Node> op::v3::Interpolate::clone_with_new_inputs(const OutputVector& new_args) const
+shared_ptr<Node> op::v4::Interpolate::clone_with_new_inputs(const OutputVector& new_args) const
 {
     check_new_args_count(this, new_args);
-    return make_shared<op::v3::Interpolate>(new_args.at(0), new_args.at(1), m_attrs);
+    if (new_args.size() <= 3)
+    {
+        return make_shared<op::v4::Interpolate>(
+            new_args.at(0), new_args.at(1), new_args.at(2), m_attrs);
+    }
+    return make_shared<op::v4::Interpolate>(
+        new_args.at(0), new_args.at(1), new_args.at(2), new_args.at(3), m_attrs);
+}
+
+namespace
+{
+    static constexpr std::size_t data_port = 0;
+    static constexpr std::size_t target_shape_port = 1;
+    static constexpr std::size_t scales_port = 2;
+    static constexpr std::size_t axes_port = 3;
+    static constexpr std::size_t max_num_of_ports = 4;
+
+    using size_t_vector = std::vector<std::size_t>;
```

Comment:
Do we really need this using (and next one)?

I don't see any win from it.

Reply:
Fixed.